### ISSUE TITLE
Setting "For" to any value above 0m for an alert based on a Cloudwatch query always results in "Pending"

### ISSUE THREAD

**Issue by driskillb43** [grafana/grafana#40391]
**What happened**:

We had a Grafana alert set such that if write capacity for one of our DynamoDB tables went above 10,000 for more than 12 hours it would immediately alert us. However, the alert failed to trigger when one of the tables was above 10,000 for a couple of days because the alert was forever in "Pending" state during that period instead of "Alerting". Based on testing, it seems that any value in the "For" field above 0m will result in the alert never firing as CloudWatch queries intermittently return "null" in the dataset which results in the alarm thinking it should not fire. Setting the period in the CloudWatch expression to the correct value (be it 1 minute, 5 minutes, etc.) does not appear to fix the issue however.

**What you expected to happen**:

The alert should have fired and only been in "Pending" for max 5 minutes (the for was set to 5m).

**How to reproduce it (as minimally and precisely as possible)**:

See video here: ![grafana](https://user-images.githubusercontent.com/5632141/137146876-5a6bb95e-8ae4-473e-88a8-bf261227f141.gif)

1. Create a DynamoDB table and set the write capacity to 5

2. Create a new Grafana dashboard and new Grafana chart with the following query:
Query Mode: CloudWatch Metrics
Region: <region>
Namespace: AWS/DynamoDB
Metric Name: ProvisionedWriteCapacityUnits
Stats: Average
TableName = <table>
Period = 5m

3. Create an alert
Name: <Name>
Evaluate every: 5m
For: 1m
When: min() OF query(A, 10m, now) IS ABOVE 3

4. Click "Test Rule" and you will see that it is pending
5. Save the dashboard and the metrics and notice that it will be pending forever
6. Change the For to 0m and notice it switches to alerting and works properly now.

**Environment**:
- Grafana version: 7.5.10 (also reproducible in [8.2.1)]
- Data source type & version: AWS CloudWatch
- OS Grafana is installed on: Provided Docker Container
- User OS & Browser: MacOS 11.5.2 and Chrome

**Comment by zuchka** [comment]
Hi @driskillb43 

Double-checking here: it looks like you are using the legacy alerting alerting system and not the newer Unified Alerting platform that shipped with Grafana 8 (under feature toggle?)

https://grafana.com/docs/grafana/latest/alerting/unified-alerting/

**Comment by driskillb43** [comment]
Hi @zuchka,

Correct, I am still using 7.5.10. I did boot up 8.2.1 but was still using the legacy alerting in it. I have not re-tested with the new Unified Alerting. 

Also, after spending some more time, I was able to get the "for" value to work but it required setting the period correctly explicitly (looking in CloudWatch and determining whether the data was every 1 minute or 5 minutes, etc.) so that no nulls were being returned in the dataset and then setting the alert query to have now-<period> at the end. I then had to wait the "for" amount after setting up the alert otherwise clicking "Test Alert" just continued to show pending until that amount time had passed.

This does not appear to necessarily be a bug with the "for" since it does work after some fiddling however I think the auto period setting for CloudWatch is not working as intended and/or it isn't clear that a null data point resets the alert in regards to the "for" setting.
